README.md
# homebridge-bme680

EXPERIMENTAL fork from https://github.com/rxseger/homebridge-bme280 providing initial bme680 support for homebridge.

Uses 
* https://github.com/trho/bsec_bme680_linux (fork from https://github.com/alexh-name/bsec_bme680_linux with json output) which relies on the closed source BME680 driver by Bosch [BME680_driver](https://github.com/BoschSensortec/BME680_driver) to calculate IAQ (Indoor Air Quality)
* https://github.com/jorisvervuurt/jvsbme680
* The original IAQ value (0-500) is shown as a fake VOCDensity characteristic without available history

Tested with Pimoroni's BME680 breakout board (https://shop.pimoroni.com/collections/electronics/products/bme680) 

Limitations:
* Only supports storage "fs", yet
* i2c address is not configurable, yet 
* Since it is based on the [fakegato-history](https://github.com/simont77/fakegato-history) library the accessory only shows up in the Eve app and not in Apple Home.

Uses heuristic factor to convert between Bosch BSEC IAQ (Indoor Air Quality) and Eve Co2 ppm:

const FACTOR_BSEC_IAQ_EVE_PPM = (CO2_MAX_VALUE - PPM_OFFSET) / BSEC_IAQ_MAX;

Concrete values:
const FACTOR_BSEC_IAQ_EVE_PPM = (3000 - 450) / 500;

[Bosch BME680](https://www.bosch-sensortec.com/bst/products/all_products/bme680)
temperature/humidity/barometric pressure sensor service plugin for [Homebridge](https://github.com/nfarina/homebridge).

[![NPM Downloads](https://img.shields.io/npm/dm/homebridge-bme280.svg?style=flat)](https://npmjs.org/package/homebridge-bme280)

* Display of temperature, humidity and Barometric Pressure from a BME280 connected to a RaspberryPI.
* Support the graphing feature of the Eve app for trends

Uses [bme280-sensor](https://www.npmjs.com/package/bme280-sensor)

# Build Instructions

1. Init
npm install
2. Compile Typescript
npm compile

TODO

## Installation

Untested!

1.	Install Homebridge using `npm install -g homebridge`
2.	Install this plugin `npm install -g trho/homebridge-bme680`
3.  Compile bsec_bme680 executable by following https://github.com/trho/bsec_bme680_linux 
4.  *Copy bsec_bme680 and bsec_iaq.config to homebridge storagePath (e.g. ~./homebridge)*
5.	Update your configuration file - see below for an example

Connect the BME680 chip to the I2C bus

## Configuration
* `accessory`: "BME680"
* `name`: descriptive name
* `name_temperature` (optional): descriptive name for the temperature sensor
* `name_humidity` (optional): descriptive name for the humidity sensor
* `name_air_quality` (optional): descriptive name for the air quality sensor
* `refresh`: Optional, time interval for refreshing data in seconds, defaults to 60 seconds.
* `options`: options for [bme280-sensor](https://www.npmjs.com/package/bme280-sensor)
* `spreadsheetId` ( optional ): Log data to a google sheet, this is part of the URL of your spreadsheet.  ie the spreadsheet ID in the URL https://docs.google.com/spreadsheets/d/abc1234567/edit#gid=0 is "abc1234567". (TODO: not implemented, yet)


Example configuration:

```json
    "accessories": [
        {
            "accessory": "BME680",
            "name": "Sensor",
            "name_temperature": "Temperature",
            "name_humidity": "Humidity",
            "useBsecLib": true,
            "bsecTempCompensation": -0.35,
            "bsecHumidityCompensation": 4.0,
            "name_air_quality": "Air Quality",
            "useBsecLib": true
        }
    ]
```

The option "useBsecLib" only works if you placed compiled bsec_bme680 in homebridge storagePath.
By omitting this option this plugin relies on jvsbme680's sensor readings. In this case IAQ is just a dummy value.
There is an advantage using this simpler solution:
The temperature and humidity readings are much more precise in this case and no callibration & compensation is necessary in this case.
Once the compensation is this configurable I will elaborate this further.
For now see kvaruni's comment on callibration & compensation for gas readings:
https://forums.pimoroni.com/t/bme680-observed-gas-ohms-readings/6608/5


This plugin creates three services: TemperatureSensor, HumiditySensor and AirQualitySensor.

## Optional - Enable access to Google Sheets to log data

TODO: not implemented, yet

This presumes you already have a google account, and have access to google drive/sheets already

Step 1: Turn on the Drive API
a. Use this wizard ( https://console.developers.google.com/start/api?id=sheets.googleapis.com )
to create or select a project in the Google Developers Console and automatically turn on the API. Click Continue, then Go to credentials.

b. On the Add credentials to your project page, click the Cancel button.

c. At the top of the page, select the OAuth consent screen tab. Select an Email address, enter a Product name if not already set, and click the Save button.  I used 'Sheets Data Logger'

d. Select the Credentials tab, click the Create credentials button and select OAuth client ID.

e. Select the application type Other, enter the name "Drive API Quickstart", and click the Create button.

f. Click OK to dismiss the resulting dialog.

g. Click the file_download (Download JSON) button to the right of the client ID.

h. Move this file to your .homebridge and rename it logger_client_secret.json.

Step 2: Authorize your computer to access your Drive Account

a. Change to the directory where the plugin is installed i.e.

cd /usr/lib/node_modules/homebridge-mcuiot/node_modules/mcuiot-logger

b. Run the authorization module

node quickstart.js

c. Browse to the provided URL in your web browser.

If you are not already logged into your Google account, you will be prompted to log in. If you are logged into multiple Google accounts, you will be asked to select one account to use for the authorization.

d. Click the Accept button.

e. Copy the code you're given, paste it into the command-line prompt, and press Enter.

## See also

* [homebridge-ds18b20](https://www.npmjs.com/package/homebridge-ds18b20)
* [homebridge-dht-sensor](https://www.npmjs.com/package/homebridge-dht-sensor)
* [homebridge-dht](https://www.npmjs.com/package/homebridge-dht)
* [bsec_bme680_linux fork] https://github.com/trho/bsec_bme680_linux
* https://github.com/rxseger/homebridge-bme280
* https://github.com/jorisvervuurt/jvsbme680


## Credits
* 
* NorthernMan54 - Barometric Pressure and Device Polling
* simont77 - History Service
* alexh-name - C wrapper for BME680 sensor BSEC library
* jorisvervuurt - Node.js module for Pimoroni's BME680 Breakout

## License

MIT


History.md
### v0.0.4
* temperature and humidity compensation configurable in config.json

### v0.0.3
* fixed history update scheduling

### v0.0.2
* Reduced characteristic service updates (temperature, humidity, IAQ, pressure) interval to 1 minute and history update interval to 10 min
* Calculate mean temperature, humidity, IAQ, pressure value from last 20 measurement in a minute interval 
* Added temperature and humidity compensation when reading IAQ values with heated sensor with bsec_bme680 library
    * see also kvaruni's comment on callibration & compensation:
    * https://forums.pimoroni.com/t/bme680-observed-gas-ohms-readings/6608/5

### v0.0.1
* Initial version, forked from https://github.com/rxseger/homebridge-bme280

src/sensor.ts
'use strict';
import { initCustomCharacteristic } from "./CustomCharacteristic";
import { EventEmitter } from "events";
import { spawn, exec, ChildProcess } from "child_process";
import { inherits } from "util";
const path = require('path');

const bme680_sensor = require('jvsbme680');
var logger = require("mcuiot-logger").logger;
const moment = require('moment');
var os = require("os");
var hostname = os.hostname();


let Service: HAPNodeJS.Service;
let Characteristic: HAPNodeJS.Characteristic;
var CustomCharacteristic;
var FakeGatoHistoryService;

let EveAirQualityPpmCharacteristic;
let EveAirQualityUnknownCharacteristic;

// TODO: make configurable
const REFRESH_TIME_IN_MINUTES = 10;

const CO2_MAX_VALUE = 3000; // I am assuming that iaq 500 ~ 3000 ppm
const BSEC_IAQ_MAX = 500;
const PPM_OFFSET = 450;

let homebridgeRef;

// Heuristic factor to convert between Bosch BSEC IAQ (Indoor Air Quality) and Eve Co2 ppm.
const FACTOR_BSEC_IAQ_EVE_PPM = (CO2_MAX_VALUE - PPM_OFFSET) / BSEC_IAQ_MAX;

module.exports = (homebridge) => {
  homebridgeRef = homebridge;
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  CustomCharacteristic = initCustomCharacteristic(homebridge);
  FakeGatoHistoryService = require('fakegato-history')(homebridge);

  homebridge.registerAccessory('homebridge-bme680', 'BME680', BME680Plugin);


  // see https://github.com/skrollme/homebridge-eveatmo/issues/1
  EveAirQualityPpmCharacteristic = function () {
    Characteristic.call(this, 'Air Quality in ppm', 'E863F10B-079E-48FF-8F27-9C2605A29F52');
    this.setProps({
      format: Characteristic.Formats.FLOAT,
      maxValue: CO2_MAX_VALUE,
      minValue: 450,
      minStep: 1,
      perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
    });
    this.value = this.getDefaultValue();
  };

  inherits(EveAirQualityPpmCharacteristic, Characteristic);

  (<any>EveAirQualityPpmCharacteristic).UUID = 'E863F10B-079E-48FF-8F27-9C2605A29F52';


  EveAirQualityUnknownCharacteristic = function () {
    Characteristic.call(this, 'Air Quality in ppm', 'E863F132-079E-48FF-8F27-9C2605A29F52');
    this.setProps({
      format: Characteristic.Formats.FLOAT,
      maxValue: CO2_MAX_VALUE,
      minValue: 450,
      minStep: 1,
      perms: [Characteristic.Perms.READ, Characteristic.Perms.NOTIFY]
    });
    this.value = this.getDefaultValue();
  };

  inherits(EveAirQualityUnknownCharacteristic, Characteristic);
  (<any>EveAirQualityUnknownCharacteristic).UUID = 'E863F132-079E-48FF-8F27-9C2605A29F52';
}

class BME680Plugin {


  toPpm(gas_resistance: number): number {
    // dummy implementation
    return PPM_OFFSET + (CO2_MAX_VALUE - PPM_OFFSET) * gas_resistance / 1000000;
  }
  
  // temperature and humidity compensation when reading IAQ values with heated sensor:
// see also kvaruni's comment on callibration & compensation:
// https://forums.pimoroni.com/t/bme680-observed-gas-ohms-readings/6608/5
// TODO: make configurable:
  bsecTempCompensation = -0.35; // or 0.7 ?
  bsecHumidityCompensation = 4.0;

  useBsecLib: boolean;
  iaqProcess: ChildProcess;
  loggingService: any;
  airQualitySensor: any; //AirQualitySensor;
  humidityService: any;//HAPNodeJS.Service.HumiditySensor;
  temperatureService: any;
  informationService: any;
  sensor: BME680;
  //  data: {};
  logger: any;
  log_event_counter: number;
  name: string;
  name_temperature: string;
  name_humidity: string;
  name_air_quality: string;
  refresh: number;
  options: any;
  spreadsheetId: any;

  iaqData: IAQResult;

  polling = false;

  constructor(private log, config) {
    this.log = log;
    this.name = config.name;
    this.useBsecLib = config.useBsecLib || false;
    this.bsecTempCompensation = config.bsecTempCompensation || 0;
    this.bsecHumidityCompensation = config.bsecHumidityCompensation || 0;
    if(config.bsecTempCompensation) {
      this.log(`Compensations: ${this.bsecTempCompensation} C, ${this.bsecHumidityCompensation} %RH`);
    }

    this.name_temperature = config.name_temperature || this.name;
    this.name_humidity = config.name_humidity || this.name;
    this.name_air_quality = config.name_air_quality || this.name;
    this.log("temp & hum names: " +
      this.name_temperature + "," +
      this.name_humidity);

    this.refresh = config['refresh'] || REFRESH_TIME_IN_MINUTES * 60;
    this.options = config.options || {};
    this.spreadsheetId = config['spreadsheetId'];
    if (this.spreadsheetId) {
      this.log_event_counter = 59;
      this.logger = new logger(this.spreadsheetId);
    }

    if ('i2cBusNo' in this.options) this.options.i2cBusNo = parseInt(this.options.i2cBusNo);
    if ('i2cAddress' in this.options) this.options.i2cAddress = parseInt(this.options.i2cAddress);
    this.log(`bme680 sensor options: ${JSON.stringify(this.options)}`);

    try {
      // Creates an instance of `BME680`.
      // uses default i2c address  [i2cAddress='0x76'] See I2C address (see `Constants.I2CAddress` for valid values)
      // Constants.I2CAddress.I2CA_PRIMARY
      this.log("init bme680 sensor ...");
      // TODO: consider this.options.i2cAddress
      if (!this.useBsecLib) {
        this.sensor = new bme680_sensor.BME680();
      }
      this.log("initialized bme680 sensor")
    } catch (ex) {
      this.log("bme680 initialization failed:", ex);
    }
    this.updateHistoryAndCurrentServices.bind(this);

    this.informationService = new Service.AccessoryInformation(undefined, undefined, undefined);

    this.informationService
      .setCharacteristic(Characteristic.Manufacturer, "Bosch")
      .setCharacteristic(Characteristic.Model, "RPI-BME680")
      .setCharacteristic(Characteristic.SerialNumber, hostname + "-" + hostname)
      .setCharacteristic(Characteristic.FirmwareRevision, require('./../package.json').version);

    this.temperatureService = new Service.TemperatureSensor(this.name_temperature, /*TRHO*/undefined, undefined);

    this.temperatureService
      .getCharacteristic(Characteristic.CurrentTemperature)
      .setProps({
        minValue: -100,
        maxValue: 100
      });

    this.temperatureService
      .addCharacteristic(CustomCharacteristic.AtmosphericPressureLevel);

    this.humidityService = new Service.HumiditySensor(this.name_humidity, undefined, undefined);

    this.airQualitySensor = new Service.AirQualitySensor(this.name_air_quality, undefined, undefined);

    this.airQualitySensor
      .addCharacteristic(EveAirQualityPpmCharacteristic);

    this.airQualitySensor
      .addCharacteristic(EveAirQualityUnknownCharacteristic);

    // Fake VOCDensity sensor used for showing BSEC IAQ value (0-500)
    // There is no history support for this value
    this.airQualitySensor
      .addCharacteristic(Characteristic.VOCDensity);
    this.airQualitySensor
      .getCharacteristic(Characteristic.VOCDensity)
      .setProps({
        minValue: 0,
        maxValue: 500
      });

    this.airQualitySensor.log = this.log;

    this.temperatureService.log = this.log;
    this.loggingService = new FakeGatoHistoryService("room", this.temperatureService,
      {
        storage: 'fs',
        disableTimer: true, // timer is not triggered; therefore disabling timer and relying own interval updates
        filename: this.name + ".json"
      });
    if (this.useBsecLib) {
      this.spawnIAQProcess();
    } else {
      this.ensurePolling();
    }

  }



  private ensurePolling() {
    if (this.polling) {
      return;
    }
    this.polling = true;
    //this.devicePolling();

    this.scheduleUpdate();
  }

  private scheduleUpdate() {
    let now = new Date();
    let firstPollAlignedWithHour = this.refresh - (now.getMinutes() * 60 + now.getSeconds()) % (this.refresh);
    this.log(`scheduled history update in ${Math.floor(firstPollAlignedWithHour / 60)} min, ${firstPollAlignedWithHour % 60} sec`);

    // update sensors, but not history:
    this.updateHistoryAndCurrentServices(false);
    setTimeout(() => {
      this.updateHistoryAndCurrentServices();
      this.scheduleUpdate();
    }, firstPollAlignedWithHour * 1000);
  }

  read(): Promise<RoomEvent> {
    if (this.iaqData) {
      return Promise.resolve(
        this.bsecResultsToEvent());
    } else {
      return this.sensor.read().then(data => {
        debug("Polling BME680: " + new Date(Date.now()).toLocaleString());
        debug(data);
        return {
          time: moment().unix(),
          temp: roundInt(data.temperature),
          pressure: roundInt(data.pressure),
          humidity: roundInt(data.humidity),
          ppm: this.toPpm(data.gasResistance) // heuristic
        };
      });
    }
  }

  private bsecResultsToEvent(): RoomEvent {

    return {
      time: moment().unix(),
      temp: roundInt(this.iaqData.raw_temperature + this.bsecTempCompensation),
      humidity: roundInt(this.iaqData.raw_humidity + this.bsecHumidityCompensation),
      pressure: roundInt(this.iaqData.pressure),
      ppm: PPM_OFFSET + this.iaqData.iaq * FACTOR_BSEC_IAQ_EVE_PPM
    };
  }

  updateHistoryAndCurrentServices(updateHistory = true) {

    this.read().then((event) => {

      if (updateHistory) {
        if (this.aggMeasurements) {
          this.log(this.aggMeasurements);
          this.aggMeasurements.time = moment().unix();
          this.loggingService.addEntry(this.aggMeasurements);
        } else {
          this.log(event);
          this.loggingService.addEntry(event);
        }

        if (this.spreadsheetId) {
          this.log_event_counter = this.log_event_counter + 1;
          if (this.log_event_counter > 59) {
            this.logger.storeBME(this.name, 0, event.temp, event.humidity, event.pressure);
            this.log_event_counter = 0;
          }
        }
      }
      let qualityDescription;
      let homeKitQuality;
      let ppm = event.ppm;

      // if (iaq >= 0 && iaq < 50) {
      //   homeKitQuality = 1;
      //   qualityDescription = "good";
      // } else if (iaq >= 50 && iaq < 100) {
      //   homeKitQuality = 2;
      //   qualityDescription = "average";
      // } else if (iaq >= 100 && iaq < 150) {
      //   homeKitQuality = 3;
      //   qualityDescription = "little bad";
      // } else if (iaq >= 150 && iaq < 200) {
      //   homeKitQuality = 4;
      //   qualityDescription = "bad";
      // } else if (iaq >= 200 && iaq < 300) {
      //   homeKitQuality = 4;
      //   qualityDescription = "worse";
      // } else if (iaq >= 300) {
      //   homeKitQuality = 5;
      //   qualityDescription = "very bad";
      // }

      if (ppm >= 450 && ppm < 700) {
        homeKitQuality = 1;
        qualityDescription = "good";
      } else if (ppm >= 700 && ppm < 1100) {
        homeKitQuality = 2;
        qualityDescription = "average";
      } else if (ppm >= 1100 && ppm < 1600) {
        homeKitQuality = 3;
        qualityDescription = "little bad";
      } else if (ppm >= 1600 && ppm < 2100) {
        homeKitQuality = 4;
        qualityDescription = "bad";
      } else if (ppm >= 2100) {
        homeKitQuality = 5;
      }

      this.log(`iaq: ${homeKitQuality} ${qualityDescription}`)
      if (this.aggMeasurements) {
        this.updateSensors(this.aggMeasurements);
      } else {
        this.updateSensors(event);
      }

      // Optional
      if (this.iaqData) {
        this.airQualitySensor
          .setCharacteristic(Characteristic.VOCDensity, roundInt(this.iaqData.iaq));
      }

      // Characteristic.AirQuality.UNKNOWN = 0;
      // Characteristic.AirQuality.EXCELLENT = 1;
      // Characteristic.AirQuality.GOOD = 2;
      // Characteristic.AirQuality.FAIR = 3;
      // Characteristic.AirQuality.INFERIOR = 4;
      // Characteristic.AirQuality.POOR = 5;
      this.airQualitySensor
        .setCharacteristic(Characteristic.AirQuality, homeKitQuality); // 1(EXCELLENT) 2 (GOOD)  3 (FAIR) 4 (INFERIOR ) 5 (POOR)
    })
      .catch(err => {
        this.log(`BME read error: ${err}`);
        debug(err.stack);
        if (this.spreadsheetId) {
          this.logger.storeBME(this.name, 1, -999, -999, -999);
        }

      });
  }

  MAX_MEASUREMENT_COUNT = 20; // collect data for 1 min (3*20 sec)

  aggMeasurements: RoomEvent;
  sumsMeasurements: RoomEvent;

  private updateSensors(event: RoomEvent) {

    this.temperatureService
      .setCharacteristic(Characteristic.CurrentTemperature, event.temp);
    this.temperatureService
      .setCharacteristic(CustomCharacteristic.AtmosphericPressureLevel, event.pressure);
    this.humidityService
      .setCharacteristic(Characteristic.CurrentRelativeHumidity, event.humidity);
    this.airQualitySensor
      .setCharacteristic(EveAirQualityPpmCharacteristic, event.ppm);

  }

  private aggregateMeasurements(event: RoomEvent) {
    if (!this.sumsMeasurements) {
      this.sumsMeasurements = event;
    } else {
      this.sumsMeasurements.temp += event.temp;
      this.sumsMeasurements.humidity += event.humidity;
      this.sumsMeasurements.pressure += event.pressure;
      this.sumsMeasurements.ppm += event.ppm;
    }
    if (this.readCounter % this.MAX_MEASUREMENT_COUNT == 0) {  // log and push aggregated values every ~ 30 sec
      this.sumsMeasurements.time = event.time;
      this.sumsMeasurements.temp = roundInt(this.sumsMeasurements.temp / this.MAX_MEASUREMENT_COUNT);
      this.sumsMeasurements.humidity = roundInt(this.sumsMeasurements.humidity / this.MAX_MEASUREMENT_COUNT);
      this.sumsMeasurements.pressure = roundInt(this.sumsMeasurements.pressure / this.MAX_MEASUREMENT_COUNT);
      this.sumsMeasurements.ppm = roundInt(this.sumsMeasurements.ppm / this.MAX_MEASUREMENT_COUNT);
      this.aggMeasurements = this.sumsMeasurements;

      let agg = this.aggMeasurements;
      this.sumsMeasurements = undefined;

      this.updateSensors(this.aggMeasurements);
      this.log(`last: ${agg.temp} (raw: ${this.iaqData.raw_temperature}, comp: ${roundInt(this.iaqData.raw_temperature + this.bsecTempCompensation)}) C, ${agg.humidity} (raw: ${this.iaqData.raw_humidity}, comp: ${roundInt(this.iaqData.raw_humidity + this.bsecHumidityCompensation)}) %RH, ${agg.ppm} ppm, IAQ: ${this.iaqData.iaq} IAQ, IAQ accuray:  ${this.iaqData.iaq_accuracy}`);
    }


  }

  getServices() {
    return [this.informationService, this.temperatureService, this.humidityService, this.airQualitySensor, this.loggingService]
  }


  /**
   * 
   * IAQ Index Air Quality
      0 – 50 good10
      51 – 100 average
      101 – 150 little bad
      151 – 200 bad
      201 – 300 worse
      301 – 500 very bad
  
   *
   * Using Eve Room, the measured gas concentration of VOCs is converted into equivalent
    CO2 concentration that is easy to read. The Eve app displays current and historical
    measurements in ppm. Values are displayed in different colors and divided into
    “Excellent” (450-700 ppm), “Good” (700-1100 ppm), “Acceptable“ (1100-1600 ppm),
    “Moderate” (1600-2100 ppm), and “Poor” (above 2100 ppm). Depending on the
    ventilation and number of people in the room, users should try to keep air quality at
    an “Excellent” or “Good” level.
   * 
   * @returns 
   * @memberof BME680Plugin
   */

  readCounter = 0;

  spawnIAQProcess() {
    if (this.iaqProcess) {
      return;
    }

    this.log(process.cwd())

    // ensure that bsec_iaq.config is located in storagePath (e.g. ~./homebridge) and is writable ( for bsec_iaq.state):
    this.iaqProcess = spawn(homebridgeRef.user.storagePath() + '/bsec_bme680', [],
      {
        cwd: homebridgeRef.user.storagePath(),
        env: process.env
      }
    );
    this.iaqProcess.stderr.setEncoding('utf8');
    this.iaqProcess.stdin.setDefaultEncoding('utf8');
    this.iaqProcess.stdout.setEncoding('utf8');

    this.iaqProcess.once('error', (error) => {
      this.cleanup();
      this.log("ERROR: bsec_bme680 failed");
      this.log(error);
    });

    this.iaqProcess.stderr.once('data', (data) => {
      this.cleanup();
      this.log("ERROR: bsec_bme680 does not run without errors, ensure that bsec_iaq.config is in the homebridge storagePath (e.g. ~./homebridge) and the directory is writable");
      this.log(data);
    });

    this.iaqProcess.stdout.on('data', (data: string) => {
      try {
        this.iaqData = JSON.parse(data);
        this.readCounter++;

        const event = this.bsecResultsToEvent();
        this.aggregateMeasurements(event);

        this.ensurePolling();
      } catch (e) {
        this.log("unable to parse iaq result: " + e);
      }
    });

    this.iaqProcess.once('exit', () => {
      this.cleanup();
    });
  }

  /**
   * Cleanup after the iaq process has exited.
   * 
   */
  private cleanup() {
    if (this.iaqProcess) {
      this.iaqProcess.removeAllListeners();
      this.iaqProcess = null;
    }
  }

}

function roundInt(string) {
  return Math.round(parseFloat(string) * 100) / 100;
}


function debug(val: any) {
  global.console.debug(val);
}

interface IAQResult {
  iaq: number;
  iaq_accuracy: number;
  raw_temperature: number;
  raw_humidity: number;
  temperature: number;
  humidity: number;
  pressure: number;
  gas_pressure: number;
  bsec_status: number;
}
interface RoomEvent {
  time: number;
  temp: number;
  pressure: number;
  humidity: number;
  ppm: number;
}

